### DISCUSSION TITLE
Make Granite into a development framwork

### DISCUSSION THREAD

**Discussion by alainm23** [elementary/granite#733] [Ideas]
At the beginning, elementary OS set itself apart by creating beautiful applications for Linux, and AppCenter is an example that the goal was (partially) achieved. I propose that Granite evolves into a development framework that makes it easier to create apps for the ecosystem.

### Use Cases
- [libadwaita](https://gitlab.gnome.org/GNOME/libadwaita): GNOME uses libadwaita to build its core apps and community apps.
- [libhelium](https://github.com/tau-OS/libhelium): This project also started developing its own framework.
- [libcosmic](https://chat.openai.com/): By Pop!_OS for its new desktop environment.

### Advantages of transforming Granite into a framework

1. **Full control over the interface:** Developers would have greater flexibility to generate dynamic and highly customized interfaces directly from code, allowing real-time adjustments and more control over the design.

2. **Structure and organization:** Providing a predefined architecture would help developers better organize their code and follow good development practices.

3. **Reusable components**: Include more pre-built widgets and components that developers can use and customize, saving time in UI creation.

4. **Modularity and extensibility:** Allow developers to add additional functionalities through modules and plugins, making the framework more flexible and adaptable to various needs.

5. **Project templates:** Tools to quickly start projects with a predefined structure.

6. **Services:** Implement a service system that allows developers to separate business logic from the UI, facilitating the reuse of common logic (such as API handling, local storage, etc.) throughout the application.

7. **Centralized state management:** Incorporate a solution to manage the application's state efficiently and reactively, facilitating data exchange between components or windows without relying on manual solutions, ensuring consistency in the UI.

8. **Window management system:** Include a window and navigation manager to simplify window lifecycle management, enabling efficient and streamlined handling of windows.

### Conclusion
Granite has the potential to evolve into a full-fledged framework. These improvements will not only benefit experienced developers but also attract new users interested in greater flexibility, scalability, and control over their applications.

**Comment by ztefn** [comment]
Isn't this just flogging a dead horse? Granite has a long way to go just to even catch up to libadwaita at this point. What would be the reason for developers to return to Granite? Many developers have moved along to Flathub and libadwaita, including [myself](https://flathub.org/apps/com.github.ztefn.haguichi) and [yourself](https://flathub.org/apps/io.github.alainm23.planify).

I think it would make more sense for elementary OS to fully embrace libadwaita instead and rebase the stylesheet on top of Adwaita (like [Yaru](https://github.com/ubuntu/yaru) for Ubuntu). This would also provide the much requested modernization of the GTK theme.

**Comment by danirabbit** [comment]
@alainm23 Is stuff like this what you have in mind? https://github.com/elementary/granite/pull/767
